client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java

```diff
@@ -818,65 +585,73 @@ private void initCatalogAndSchema()
-        Map<String, Object> suppliedProperties = driverProperties.entrySet().stream()
-                .collect(toImmutableMap(entry -> (String) entry.getKey(), Entry::getValue));
-
-        for (String key : urlProperties.keySet()) {
-            if (suppliedProperties.containsKey(key)) {
-                throw new SQLException(format("Connection property %s is both in the URL and an argument", key));
+        Properties urlProperties = extractPropertiesFromUri(uri, restrictedOptions);
+        for (Object key : urlProperties.keySet()) {
+            if (key != SSL.getKey() && driverProperties.containsKey(key)) {
+                throw new RuntimeException(format("Connection property %s is both in the URL and an argument", key));
             }
         }
 
         Properties result = new Properties();
-        setProperties(result, suppliedProperties);
-        setProperties(result, urlProperties);
+        result.putAll(urlProperties);
+        // Order is important if i.e. driverProperties want to disable SSL
+        result.putAll(driverProperties);
+
         return result;
     }
 
-    private static void setProperties(Properties properties, Map<String, Object> values)
+    private void validateConnectionProperties(Properties connectionProperties)
     {
-        properties.putAll(values);
-    }
+        ImmutableList.Builder<RuntimeException> violations = ImmutableList.builder();
 
-    private static void validateConnectionProperties(Properties connectionProperties)
-            throws SQLException
-    {
         for (String propertyName : connectionProperties.stringPropertyNames()) {
             if (ConnectionProperties.forKey(propertyName) == null) {
-                throw new SQLException(format("Unrecognized connection property '%s'", propertyName));
+                violations.add(new IllegalArgumentException(format("Unrecognized connection property '%s'", propertyName)));
             }
         }
 
-        for (ConnectionProperty<?, ?> property : ConnectionProperties.allProperties()) {
-            property.validate(connectionProperties);
+        for (ConnectionProperty<?, ?> property : allProperties()) {
+            Optional<RuntimeException> validationError = property.validate(connectionProperties);
+            validationError.ifPresent(violations::add);
+        }
+
+        if (hasPassword() && !isUseSecureConnection()) {
+            violations.add(new IllegalStateException("TLS/SSL is required for authentication with username and password"));
+        }
+
+        List<RuntimeException> errors = violations.build();
+        if (errors.size() == 1) {
+            throw errors.get(0);
+        }
+        else if (!errors.isEmpty()) {
+            String multipleViolations = errors.stream()
+                    .map(RuntimeException::getMessage)
+                    .collect(joining("\n"));
+
+            throw new RuntimeException("Provided connection properties are invalid:\n" + multipleViolations);
```

Comment:
I'd join them without any prefix. Every error should form a correct English sentence, and we should terminate them with a dot. We shouldn't assume they'll be displayed to the user with newlines, some JDBC clients could mess up the formatting.